### PR TITLE
use string equality instead of regexps to compare constant strings

### PR DESCRIPTION
**Pull request by erik-krogh**
Regexps have previously been used to encode a set of strings, but now that we have set literals we should use those instead.  
And sometimes regexps seem to be used without any good reason (second example below).  

Consider the below examples.  
```CodeQL
str.regexpMatch("foo|bar|baz") or // is equivalent to the below, but less efficient. 
str = ["foo", "bar", "baz"]
or
str.matches("foo") or // is equivalent to the below, but less efficient. 
str = "foo"
``` 

Rewriting from a regular expression to comparison against string constants means that there are more possible join-orders for the optimizer to choose from, so we need to run performance evaluations.  

---- 

Evaluations all look good: [CPP](https://github.com/github/codeql-dca-main/tree/data/PR-9182-0-cpp/reports), [C#](https://github.com/github/codeql-dca-main/tree/data/PR-9182-1-csharp/reports), [JavaScript](https://github.com/github/codeql-dca-main/tree/data/PR-9182-3-javascript/reports), [Java](https://github.com/github/codeql-dca-main/tree/data/PR-9182-2-java/reports), [Python](https://github.com/github/codeql-dca-main/tree/data/PR-9182-4-python/reports), [Ruby](https://github.com/github/codeql-dca-main/tree/data/PR-9182-5-ruby/reports).    
I had to do a bunch of retries in C++ before that turned out OK, but I think that's due to big variations in DB creation time  
(DB caching in DCA :pray:).  

---- 

Inspired by this comment: https://github.com/github/codeql/discussions/5160#discussioncomment-2762679

### CODE REVIEW

File: ql/ql/src/queries/style/UseStringComp.ql [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+  call.getBase().getType().getASuperType*().(PrimitiveType).getName() = "string" and
+  (
+    call.getMemberName() = "regexpMatch" and
+    call.getNumberOfArguments() = 1 and
+    call.getArgument(0).(String).getValue().regexpMatch("([a-zA-Z0-9]+\\|)*[a-zA-Z0-9]+")
+    or
+    exists(string reg | call.getMemberName() = "matches" |
+      call.getNumberOfArguments() = 1 and
+      reg = call.getArgument(0).(String).getValue() and
+      not reg.regexpMatch(".*%.*")
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply ql/ql/src/queries/style/UseStringComp.ql]: > However I also note that most cases with underscores seem to be bugs where an escaped underscore is intended.  

I found two uses of underscores, both were bugs.  
I've fixed them.

File: java/ql/lib/semmle/code/java/Type.qll [hunk 2]
```diff
@@ -1261,9 +1263,9 @@
 class FloatingPointType extends Type {
   FloatingPointType() {
     exists(string name |
-      name = this.(PrimitiveType).getName() or name = this.(BoxedType).getPrimitiveType().getName()
+      name = [this.(PrimitiveType).getName(), this.(BoxedType).getPrimitiveType().getName()]
     |
-      name.regexpMatch("float|double")
+      name = ["float|double"]
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply java/ql/lib/semmle/code/java/Type.qll]: TP :tada:

**Review by hvitved — APPROVED** [review]
C# and Ruby 👍

**Review by aschackmull — APPROVED** [review]
Java LGTM

**Review by jketema — APPROVED** [review]
C++ 👍

**Review by esbena — APPROVED** [review]
JS 👍🏻

**Review by tausbn — APPROVED** [review]
Python/QL :+1:

**Pull request merged**